docs/uenv-compilation-spack.md

```diff
@@ -0,0 +1,163 @@
+
+```
+cp2k@2024.1 +cuda cuda_arch=80
+```
+
+This [spec] defines a build of [CP2K] version `2024.1` with CUDA acceleration. All other dependencies and features are defined by the default values in the [CP2K Spack package].
+
+## Set uenv as upstream Spack instance
+
+Activate the uenv. Here we assume the uenv described above is called `cp2k/2024.1` and it is already deployed:
+
+```bash
+uenv image pull cp2k/2024.1
+uenv start cp2k/2024.1
+```
+
+With the uenv active, we can now tell our local [Spack] instance to use the uenv as an upstream [Spack] instance (see [Chaining Spack Installations] for more details):
+
+```bash
+export SPACK_SYSTEM_CONFIG_PATH=/user-environment/config/
+```
+
+!!! note
+    We assumed here that the uenv is mounted in the standard location `/user-environment/`. If it is mounted in a non-standard location, adjust the previous command accordingly.
+
+## Building your own version 
+
+Let's assume we want to have a version of [CP2K] what uses the COSMA library for communication-optimal matrix multiplication with GPU-aware MPI:
+
+```
+cp2k@2024.1 +cuda cuda_arch=80 +cosma ^cosma +gpu_direct
+```
+
+COSMA is not availabe in the uenv described above since it is not a dependency needed by `cp2k@2024.1 +cuda cuda_arch=80`. 
+
+### Spack Environment
+
+To make things clean and reproducible, we use [Spack Environments] do describe what we want to build. To define a [Spack environment] we create the following file in a folder (hereafter referred to as `SPACK_ENV_FOLDER`):
+
+```yaml
+spack:
+  specs:
+  - cp2k@2024.1 +cosma
+  packages:
+    all:
+      prefer:
+        - +cuda cuda_arch=80
+    cosma:
+      require:
+        - +gpu_direct
+  view: false
+  concretizer:
+    unify: true
+```
+
+`packages:all:prefer` indicates that we want the `+cuda` variant active for all packages that have it. The `require` clause is stronger, and we use it for a specific package, COSMA, to make sure the `+gpu_direct` variant (GPU-aware MPI) is enabled (this is a constraint we want to enforce on the COSMA dependency). 
+
+!!! note
+    It is good practice to have a single root [spec] in an environment, and to define constraint on all packages or specific dependencies in the `packages:` field.
+
+### Building
+
+After defining the environment above, we can concretize it:
+
+```bash
+spack -e SPACK_ENV_FOLDER concretize -f
+```
+
+The result of the concretization will be printed on screen. Packages starting with `[+]` are packages that will be freshly installed in your local [Spack] instance. Packages marked ad `[e]` (external) are packages taken directly from the uenv (which we are using as upstream [Spack] instance). You should see many packages marked as `[e]`, which are being re-used from the uenv. This will greatly speed up compilation, since [Spack] will have to build only a small subset of packages.
```

Comment:
```suggestion
The result of the concretization will be printed on screen. Packages starting with `[+]` are packages that will be freshly installed in your local [Spack] instance. Packages marked as `[e]` (external) are packages taken directly from the uenv (which we are using as upstream [Spack] instance). You should see many packages marked as `[e]`, which are being re-used from the uenv. This will greatly speed up compilation, since [Spack] will have to build only a small subset of packages.
```